raiden/utils/migrations/v17_to_v18.py

```diff
@@ -0,0 +1,88 @@
+import json
+
+from raiden.exceptions import ChannelNotFound
+from raiden.storage.sqlite import SQLiteStorage
+from raiden.transfer.state import RouteState
+from raiden.utils.typing import Any, Dict
+
+SOURCE_VERSION = 17
+TARGET_VERSION = 18
+
+
+def get_token_network_by_identifier(snapshot, token_network_identifier):
+    identifiers_to_paymentnetworks = snapshot['identifiers_to_paymentnetworks']
+    for paymentnetwork in identifiers_to_paymentnetworks.values():
+        for token_network in paymentnetwork['tokennetworks']:
+            if token_network['address'] == token_network_identifier:
+                return token_network
+    return None
+
+
+def _transform_snapshot(raw_snapshot: Dict[Any, Any]):
+    """
+    This migration upgrades the object:
+    - `MediatorTransferState` such that a list of routes is added
+    to the state to be able to route a waiting transfer in case the
+    receiving node comes back online.
+    """
+    snapshot = json.loads(raw_snapshot)
+    secrethash_to_task = snapshot['payment_mapping']['secrethashes_to_task']
+    for task in secrethash_to_task.values():
+        if task['_type'] != 'raiden.transfer.state.MediatorTask':
+            continue
+
+        mediator_state = task.get('mediator_state')
+
+        # The migration should be idempotent.
+        if 'routes' in mediator_state:
+            continue
+
+        mediator_state['routes'] = []
+
+        waiting_transfer = mediator_state.get('waiting_transfer')
+        if waiting_transfer is None:
+            continue
+
+        transfer = waiting_transfer.get('transfer')
+        token_network_identifier = transfer['balance_proof']['token_network_identifier']
+        token_network = get_token_network_by_identifier(
+            snapshot,
+            token_network_identifier,
+        )
+        channel_identifier = transfer['balance_proof']['channel_identifier']
+        channel = token_network.get('channelidentifiers_to_channels').get(channel_identifier)
+        if not channel:
+            raise ChannelNotFound(
+                'Upgrading to v18 failed. '
+                f'Could not find channel with identifier {channel_identifier} '
+                'in the current chain state.',
+            )
+
+        # Only add the route for which the waiting transfer was intended.
+        # At the time of migration, we cannot re-calculate the list of routes
+        # that were originally calculcated when the transfer was being
+        # mediated so this step should be sufficient for now.
+        mediator_state['routes'] = [
```

Comment:
question: This step is also not required right? When raiden restarts it will calculate the routes again?

Reply:
Assuming `A-B-C`, and `B` currently holds a waiting transfer to C:
The idea here is that we add B-C as a route so that after the upgrade, if C becomes available, B would react to this "state change" by trying to send the transfer again to C. (see `mediator.handle_node_change_network_state`)

Reply:
I see. Thank you for the explanation.